### PR TITLE
Document oidc-agent, fedcloudclient and terraform

### PR DESCRIPTION
**Pull request by gwarf**
Document using oidc-agent, fedcloudclient and terraform altogether.
Using Dynamic DNS too.

I had to do this for myself so thought it may be nice to have some kind of of very guided tutorial/setup for starting up with a more advanced CLI usage.

### CODE REVIEW

**Comment by gwarf** [comment]
I've also some bits allowing to prepare some kind of template repository, that could also include `requirements.txt` to setup the python module, required `terraform` files and also using `terraform` an ansible inventory file, and a `requirements.yml` to install some ansible module and run it against the deployed VM.
Maybe having this as a ral repo template may be handy/useful for some?

**Comment by EGI-ILM** [comment]
I am getting 404 from https://docs.egi.eu/documentation/515

Not sure if this is related or useful, but I documented some guides here:
- https://github.com/EGI-ILM/fedcloud-terraform
- https://github.com/EGI-ILM/automated-containers

**Comment by gwarf** [comment]
The 404 is the usual one when the preview is marked as deployed but not yet reachable for real, now it's here: https://docs.egi.eu/documentation/515/users/tutorials/oidc-agent-fedcloudclient-terraform/ (it's a WIP, some content, like the terraform files, is missing).

Currently *my* position/feeling is to aim at "promoting" tools that are properly maintained by an upstream and that we are sure will stay for long (ie thinks like terraform, oidc-agent and fedcloudclient, compared to home made scripts) , but I think providing a bunch of places/repo where people could look for more advanced stuff would be nice. But drawing the line about what to document or not is always difficult, especially as some stuff may become outdated.

Concerning your repositories (nice work!), we may at least want to reference them somewhere, so that people can find them.
In the end I feel we are apparently all maintaining some personal documentation or even advanced scripts like yours, so while it's expected people will have to dive into stuff to understand/own them, there is some duplication of effort and lack of reuse and sharing, and I feel those resources may be interesting for other users.

Maybe we could at least add a section with some references to advanced tools/resources/repositories with a short description?

**Comment by EGI-ILM** [comment]
I agree, there is an inherent peril in maintaining documentation about products and tools that change rapidly, as the documentation becomes outdated and it is impossible to keep up.

I shared the links in my comment exactly for the issue you mention: checking if my work could help to avoid duplicated work. There is actually overlapping between my Terraform repository and this new tutorial.
In my case, I created the repositories to address specific requests from the community ("how can I use Terraform on EGI infrastructure" and "how can I run disposable containers with automated means on EGI infrastructure" respectively). I just wanted to give a tested, specific step-by-step guide, which is easier than going through the whole documentation of disparate sources. In the case of Terraform I gave manual steps, but also created a user-friendly script, something as simple as possible for the user to have all the work done for them (e.g. it automatically creates Terraform files with the available flavours of a specific site, etc). This will solve the current community needs, but may require some updates in future...

As you mention, maybe the best is to link these guides from the docs. This way, if something is not relevant anymore, the link can be easily removed. Or it can be kept to serve as the base for updates or new requests. I think it is important to keep track of this efforts and knowledge, otherwise it will get lost. For example, other people did also a similar exercise with Terraform in the past, but this effort is not registered anywhere.

**Comment by gwarf** [comment]
@enolfc what is your opinion on something like this PR? Worth to be included in official doc or to be kept outside?

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+package for a command-line client designed for interaction with the OpenStack
+services in the EGI infrastructure. The client can access various EGI services
+and can perform many tasks for users including managing access tokens, listing
+services, and mainly execute commands on OpenStack sites in EGI infrastructure.
+
+`fedcloudclient` can leverage [`oidc-agent`](#setting-up-oidc-agent) if it's
+installed and properly configured.
+
+`fedcloudclient` and
+[`opentackclient`](https://docs.openstack.org/python-openstackclient/latest/),
```
Comment (enolfc) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:116]:
```suggestion
[`openstackclient`](https://docs.openstack.org/python-openstackclient/latest/),
```

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+It's possible to deploy an OpenStack Virtual Machine (VM) on any of the sites
+supporting the Virtual Organisations (VO) you are a member of.
+
+Once [fedcloudclient](#installing-fedcloudclient-and-ansible) is installed it's
+possible to get information about the OIDC token accessed via
+[oidc-agent](#setting-up-oidc-agent).
+
+```shell
+# Verifying that an OS_ACCESS_TOKEN is available
+$ echo $OS_ACCESS_TOKEN
```
Comment (enolfc) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:155]: why not just using the oidc agent support in fedcloudclient?

**Review by enolfc — COMMENTED** [review]
This is quite nicely written. I like it and I think it can be useful.
Just 2 small things

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+[identified](#identifying-the-cloud-resources) and documented in a
+[`$EGI_SITE.tfvars` file](#documenting-the-cloud-resources-for-the-selected-site).
+
+#### Identifying the cloud resources
+
+Identify flavor, image, network and security groups for the site you want to
+use, using the information gathered with `fedcloudclient`.
+
+```shell
+# Selecting an image
```
Comment (sebastian-luna-valero) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:252]:
```suggestion
# Configure site and VO
$ export EGI_VO='vo.access.egi.eu'
$ export EGI_SITE='IN2P3-IRES'
# Selecting an image
```

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+$ fedcloud select image --image-specs "Name =~ 'EGI.*22'"
+# Selecting a flavor
+$ fedcloud select flavor --flavor-specs "RAM>=2096" \
+    --flavor-specs "Disk > 10" --vcpus 2
+# Identifying available networks
+$ fedcloud openstack --site "$EGI_SITE" network list
```
Comment (sebastian-luna-valero) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:258]:
```suggestion
$ fedcloud openstack network list
```

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+$ fedcloud select network --network-specs default
+# Identifying security groups
+$ fedcloud openstack --site "$EGI_SITE" security group list
```
Comment (sebastian-luna-valero) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:261]:
```suggestion
$ fedcloud openstack security group list
```

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+# Listing rules of a specific security group
+$ fedcloud openstack --site "$EGI_SITE" security group rule list default
```
Comment (sebastian-luna-valero) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:263]:
```suggestion
$ fedcloud openstack security group rule list default
```

**Review by sebastian-luna-valero — COMMENTED** [review]
This is a great tutorial, many thanks!

Minor comments from me below.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 3 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:252]: Setting up the environment is previously documented in https://docs.egi.eu/documentation/515/users/tutorials/oidc-agent-fedcloudclient-terraform/#setting-up-the-environment.
I'm pushing an update pointing to this section.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 4 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:258]: I initialy tried without but for me it doesn't work, that's why I added the `--site "$EGI_SITE"`:
Screenshot 2022-10-20 at 09 55 17

Is it working for you without the explicit `--site` parameter?

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 2 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md]: You are right, I removed this and marked exporting `OS_ACCESS_TOKEN` as an example.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 4 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:258]: For the record things like `fedcloud site env` work correctly and take into account the `EGI_SITE` and `EGI_VO` variables, it's only for `fedcloud openstack` that the `EGI_SITE` is apparently not taken into account:
Screenshot 2022-10-20 at 10 12 21

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 7]
```diff
@@ -505,6 +505,29 @@ Once this works, you can
...
+
+## Additional resources
+
+Additional resources are available, and can help with addressing different use
+cases, or be used as a source of inspiration:
+
+- [egi-qc/deployment-howtos](https://github.com/egi-qc/deployment-howtos):
+  Deployment recipes extracted from Jenkins builds for the
+  [UMD](https://go.egi.eu/umd) and [CMD](https://go.egi.eu/cmd) products
+- [EGI-ILM/fedcloud-terraform](https://github.com/EGI-ILM/fedcloud-terraform):
```
Comment (gwarf) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:525]: @EGI-ILM I've added links to your repositories in a new section about additional resources.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 4 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:258]: Discussed out of band with @sebastian-luna-valero , same behaviour for him: it doesn't  work without the `--site`.
It should in theory work like this, the documentation will be updated once the problem got identified/fixed.

Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:258]: Rejecting suggestion for now.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 6 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:263]: Rejecting suggestion for now.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 5 above)
Comment (gwarf) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:261]: Rejecting suggestion for now.

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md (hunk 4 above)
Comment (sebastian-luna-valero) [reply content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:258]: xref: https://github.com/tdviet/fedcloudclient/issues/150

File: content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md [hunk 8]
```diff
@@ -248,6 +248,12 @@ Once the [environment is properly configure](#setting-up-the-environment),
 `fedcloudclient` is used to gather information and identify flavor, image,
 network and security groups for the site you want to use.
 
+> `fedcloud openstack` currently requires an explicit `--site` parameter, this
+> will be addressed in a
+> [future fedcloud release](https://github.com/tdviet/fedcloudclient/issues/150).
+> In the meantime the `$EGI_VO` environment variable can be reused using
+> `--site "$EGI_VO"`.
```
Comment (sebastian-luna-valero) [inline content/en/users/tutorials/oidc-agent-fedcloudclient-terraform/_index.md:255]:
```suggestion
> In the meantime the `$EGI_SITE` environment variable can be reused using
> `--site "$EGI_SITE"`.
```

**Review by sebastian-luna-valero — APPROVED** [review]
Fine by me, many thanks again for putting this together!

**Review by enolfc — APPROVED** [review]
LGTM

**Pull request merged**